File: EVM.py
"""EARNED VALUE MANAGEMENT (EVM) CALCULATOR APP """

"""
This simple program will perform Earned Value Management (EVM) analysis calculations for controlling/monitoring
cost when managing projects. All the EVM formulas will be defined below. This application also assumes that the
planned % completed will always be given as a percentage. If while conducting EVM analysis and the
planned % completed is not provided "explicitly", then take the now(current time) and divide by the
overall project duration.

"""

# User to provide budget expenditures and actual project cost
BAC_request = int(
    input("Enter budget at completion $")
)
"""
BAC is the Budget At Completion or original budget of the project
"""
AC_request = int(input("Enter actual cost $"))
"""
AC is the Actual Cost or the amount of money already spent on the project
"""
# User to provide the percent complete for both planned activities and the actual percent complete for work done
planned_percent_complete = float(input("Enter Planned % Completed "))
actual_percent_complete = float(input("Enter the Actual % Completed "))

# Compute the Planned and Earned Value estimates respectfully
PV = planned_percent_complete * BAC_request
"""
Amount of money worth of work that we should have been done on the project. 
"""
EV = actual_percent_complete * BAC_request
"""
Amount of money worth of work you actually did on the project. 
"""


print("The Planned value is: $" + str(PV))
print("The Earned value is: $" + str(EV))


def costVariance(EV1):
    """
    The difference between the work done and money spent. This value
    should be positive for under budget. Negative values indicate
    under budget.
    """
    result = EV - AC_request
    if result < 0:
        print("Project is over budget")
    else:
        print("Project is under budget")

    return "The cost variance is: $" + str(result)


print(costVariance(EV))


def costPerformanceIndex(CPI):
    """
    The rate of how we are spending to actually earning on
    the project. The value should be 1 and over for projects
    under budget.
    """
    CPI = EV / AC_request
    EAC = BAC_request / CPI
    """
    Forcasting the total cost of the project at the end
    based on the current spending rate of the project. 
    """
    print("The Estimate at Completion is: $" + str(round(EAC, 2)))
    ETC = EAC - AC_request
    """
    Forcasting the amount that will be needed to complete the current
    project based on the current performance.
    """
    print("The Estimate to Completion is: $" + str(round(ETC, 2)))
    VAC = BAC_request - EAC
    """
    The difference between the original budget and new forecasted budget. This value should
    be positive for projects that may end at or under budget.
    """
    print("The Variance at Completion is: $" + str(round(VAC)))
    """The performance that needs to be met to finish the project within the budget"""
    TCPI = (BAC_request - EV) / (BAC_request - AC_request)
    """
    The performance that needs to be met to finish the project within the budget. 
    """
    print("The To-Complete Performance Index is: " + str(round(TCPI, 2)))
    if CPI >= 1:
        print("Project is under budget")
    else:
        print("Project is over budget")
    if AC_request == 0:
        raise ZeroDivisionError("Opps, division by zero detected")
    return round(CPI, 1)


print("The Cost Performance Index is: " + str(costPerformanceIndex(EV / AC_request)))


def scheduleVariance(EV1, PV1):
    """
    The difference between the amount of work we should have done vs. the amount actually done.
     This value should be positive for ahead of schedule. Negative values indicate
     behind schedule.
    """
    answer2 = EV - PV
    if answer2 < 0:
        print("Project is behind schedule")
    else:
        print("Project is ahead of schedule")
    return "The schedule variance is: $" + str(answer2)


print(scheduleVariance(EV, PV))


def schedulePerformanceIndex(EV1, PV1):
    """
    The rate of how we are meeting the project schedule. This value
    should be 1 and over for a project to be ahead of schedule.
    """
    answer3 = EV / PV
    if answer3 >= 1:
        print("Project is ahead of schedule")
    else:
        print("Project is behind schedule")
    if PV == 0:
        raise ZeroDivisionError("Opps, division by zero detected")
    return "The schedule performance index is: " + str(round(answer3, 1))
    # return answer3


print(schedulePerformanceIndex(EV, PV))

File: EVM.kt
/* The code below will perform earned value calculations for monitoring cost and scheduling for typical projects. The user will be prompted to enter certain values
 in order to complete the formulas below. */

fun main()
{
    // Prompt user to enter project budget at completion
    println("The Budget at Completion is: ")
    var budgetAtCompletion = readLine()!!
    println("The Budget at Completion is: $${budgetAtCompletion.toDouble()}")

    // Prompt user to enter planned % complete
    println("Enter planned % complete")
    var plannedPercentComplete = readLine()!!
    println("Planned % Complete is: $plannedPercentComplete ")

    // Prompt user to enter actual % complete
    println("Enter the actual % complete")
    var actualPercentComplete = readLine()!!
    println("Actual % complete is: ${actualPercentComplete.toDouble()}")

    // Prompt user to enter actual cost for the project
    println("Enter the actual cost")
    var actualCost = readLine()!!
    println("The Actual amount of money spent is: $${actualCost.toDouble()}")

    // Estimate the Planned Value(PV) and the Earned Value(EV) calculations
    var plannedValue = plannedPercentComplete.toDouble() * budgetAtCompletion.toDouble()
    var earnedValue = actualPercentComplete.toDouble() * budgetAtCompletion.toDouble()


    println("The estimated planned value is: $$plannedValue")
    println("The estimated earned value is: $$earnedValue")

    // Call the cost variance function
    var result = costVariance(earnedValue, actualCost.toDouble())
    // Check if the result is either a positive or negative variance; if negative the project is over budget
      if(result > 0) {
        println("The project is under budget!")

    } else {
        println("The project is over budget!")
    }
    println("The cost variance is: $${result.toDouble()}")

    // Estimate the cost performance index(CPI)
    var CPI = earnedValue.toDouble() / actualCost.toDouble()
    println("The cost performance index is: ${CPI}")
    // Check if the value of the CPI is greater than 1; if less than 1 the project is over budget
    if (CPI >= 1.0) {
        println("The project is under budget")

    } else {
        println("The project is over budget")
    }

    //Call the schedule variance(SV)function
    var resultSV = scheduleVariance(earnedValue.toDouble(), plannedValue.toDouble())
    // Check if the result is either a positive or negative variance; if negative project is behind schedule
     if (resultSV > 0) {
         println("The project is ahead of schedule!")
     } else {
         println("The project is behind schedule!")
     }
     println("The schedule variance is: $${resultSV.toDouble()}")

     // Estimate the schedule variance index(SPI)
      var SPI = earnedValue.toDouble() / plannedValue.toDouble()
      println("The schedule performance index is: $SPI")
    // Check if the SPI is greater than 1; if less than 1 the project is behind schedule
      if (SPI >= 1.0) {
          println("The project is ahead of schedule")
      } else {
          println("The project is behind schedule")
      }

    // Calculate the Estimate at completion(EAC) and use exception handling to check for division by zero
      var EAC = budgetAtCompletion.toDouble() / CPI.toDouble()

      if (CPI == 0.0) {
        try // throw an exception
        {
            var EAC = budgetAtCompletion.toDouble() / 0.0
            println("An error has occured.")
        }
        catch (e: ArithmeticException) { // catch and handle the exception
            println("Division by zero not allowed!")
        }
    }
     println("$$EAC is the estimate at completion")
     // Calculate the Estimate to Complete(ETC)
    var ETC = EAC.toDouble() - actualCost.toDouble()
    println("$$ETC is the Estimate to Completion")

     // Calculate the Variance of completion(VAC)
    var VAC = budgetAtCompletion.toDouble() - EAC.toDouble()
    println("$$VAC is the Variance at Completion")

      // To-Complete performance Index(TCPI)
     var TCPI = (budgetAtCompletion.toDouble() - earnedValue.toDouble()) /
            (budgetAtCompletion.toDouble() - actualCost.toDouble())


}


    // Use function declarations below to estimate the cost and schedule variances
fun costVariance(earnedValue: Double, actualCost: Double): Double
{
    return earnedValue - actualCost
}


fun scheduleVariance(earnedValue: Double, plannedValue: Double): Double {
    return earnedValue - plannedValue
}
